src/language/dafnyInstallation.ts

```diff
@@ -28,6 +28,27 @@ async function ifNullOrEmpty(a: string | null, b: () => Promise<string>): Promis
   return a === null || a === '' ? await b() : Promise.resolve(a);
 }
 
+export async function getServerApiVersion(): Promise<string | undefined> {
+  const runtimePath = getConfiguredLanguageServerRuntimePath();
+  if(runtimePath !== '') {
+    return undefined;
+  }
+
+  const [ , version ] = await getConfiguredTagAndVersion();
+  if(!/(\d)\.(\d)\.(\d)/g.test(version)) {
+    return undefined;
+  }
+  return version;
+}
+
+export function versionToNumeric(version: string | undefined): number {
+  if(version === undefined) {
+    return Number.MAX_VALUE;
```

Comment:
Aha nice hack to ensure a custom installation path is treated as the most recent Dafny version.
However, this might not always be true: If I checkout a previous installation of Dafny to test a certain behavior, it will treat it like it was the last version and your code will fail to display previous messages.
I should have thought of that earlier, but instead of checking the server's path and inferred version, could you please look into "capabilities"?
I see a message that the language server sends to VSCode at the very beginning:

```
Received response 'initialize - (0)' in 2312ms.
Result: {
    "capabilities": {
        "experimental": {},
        "workspace": {
            "workspaceFolders": {
                "supported": true,
                "changeNotifications": true
            },
            "fileOperations": {}
        }
    },
    "serverInfo": {
        "name": "DafnyLanguageServer"
    }
```

I think you should be able to detect there if the language server has the capability to send the test items. That will be more robust than testing the version (testing the version would be as robust but only if the serverInfo contained the version number of Dafny...)

Reply:
Capabilities seem useful when there's a protocol and different servers implement different parts of it. For Dafny, we have Dafny specific protocol extensions and the latest Dafny server will always implement everything in the latest protocol version, so I think a simple API version number should be sufficient.

But I agree it would be best if the server vends this API version number.